python/mscclpp/language/program.py

```diff
@@ -430,7 +489,125 @@ def print_chunk_info(self, index=0):
         print(self._get_chunk(index + self.index))
 
 
+class OrderDAG:
+    start_nodes: list["Node"]
+    last_node: Dict[Tuple[int, int], int]
+    signalling: Dict[Tuple[int, int, int], Queue]
+    waiting: Dict[Tuple[int, int, int], Queue]
+
+    def __init__(self):
+        self.start_nodes = []
+        self.last_node = {}
+        self.signalling = {}
+        self.waiting = {}
+
+    def insert_operation(self, op: "Op"):
+        node = self.Node(op)
+        rank = op.rank
+        tb = op.tb
+
+        if op.inst == Instruction.barrier:
+            for tb in op.extra.get("tb_list", None):
+                if (rank, tb) not in self.last_node:
+                    self.last_node[(rank, tb)] = node
+                    self.start_nodes.append(node)
+                else:
+                    prev_node = self.last_node[(rank, tb)]
+                    prev_node.next_nodes.append(node)
+                    node.input += 1
+                    self.last_node[(rank, tb)] = node
+
+        else:
+            if (rank, tb) not in self.last_node:
+                self.last_node[(rank, tb)] = node
+                self.start_nodes.append(node)
+            else:
+                prev_node = self.last_node[(rank, tb)]
+                prev_node.next_nodes.append(node)
+                node.input += 1
+                self.last_node[(rank, tb)] = node
+
+        if op.inst == Instruction.signal:
+            if (op.src.rank, op.dst.rank, tb) not in self.waiting or self.waiting[
+                (op.src.rank, op.dst.rank, tb)
+            ].empty():
+                if (op.src.rank, op.dst.rank, tb) not in self.signalling:
+                    self.signalling[(op.src.rank, op.dst.rank, tb)] = Queue()
+                self.signalling[(op.src.rank, op.dst.rank, tb)].put(node)
+            else:
+                waiting_node = self.waiting[(op.src.rank, op.dst.rank, tb)].get()
+                node.next_nodes.append(waiting_node)
+                waiting_node.input += 1
+
+        if op.inst == Instruction.wait:
+            if (op.src.rank, op.dst.rank, tb) not in self.signalling or self.signalling[
+                (op.src.rank, op.dst.rank, tb)
+            ].empty():
+                if (op.src.rank, op.dst.rank, tb) not in self.waiting:
+                    self.waiting[(op.src.rank, op.dst.rank, tb)] = Queue()
+                self.waiting[(op.src.rank, op.dst.rank, tb)].put(node)
+            else:
+                signalling_node = self.signalling[(op.src.rank, op.dst.rank, tb)].get()
+                signalling_node.next_nodes.append(node)
+                node.input += 1
+
+    def execute_operations(self):
```

Comment:
I think you try to implement a `topological-sorting` right? Maybe we can refer to some existing python implementation for this. I found a library for this: https://docs.python.org/3.10/library/graphlib.html. Maybe you can find smarter version